NotificationService: Crash when reading the bytes of an old alarm after a schema change

https://ankidroid.org/acra/app/1/bug/396907/report/ca2337af-89c5-4595-beb0-dc57e02a62a5

2.24.0alpha14

`SDK_INT: 33`; Xiaomi Poco F5 Pro

```
java.lang.RuntimeException: Unable to start receiver com.ichi2.anki.services.NotificationService: android.os.BadParcelableException: Parcelable protocol requires subclassing from Parcelable on class za.a0
	at android.app.ActivityThread.handleReceiver(ActivityThread.java:4501)
	at android.app.ActivityThread.-$$Nest$mhandleReceiver(Unknown Source)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2227)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:210)
	at android.os.Looper.loop(Looper.java:299)
	at android.app.ActivityThread.main(ActivityThread.java:8247)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:559)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:954)
Caused by: android.os.BadParcelableException: Parcelable protocol requires subclassing from Parcelable on class za.a0
	at android.os.Parcel.readParcelableCreatorInternal(Parcel.java:4872)
	at android.os.Parcel.readParcelableInternal(Parcel.java:4766)
	at android.os.Parcel.readParcelable(Parcel.java:4743)
	at com.ichi2.anki.reviewreminders.ReviewReminder$Creator.createFromParcel(ReviewReminder.java)
	at com.ichi2.anki.reviewreminders.ReviewReminder$Creator.createFromParcel(ReviewReminder.java)
	at android.os.Parcel.readParcelableInternal(Parcel.java:4775)
	at android.os.Parcel.readValue(Parcel.java:4532)
	at android.os.Parcel.readValue(Parcel.java:4312)
	at android.os.Parcel.-$$Nest$mreadValue(Unknown Source)
	at android.os.Parcel$LazyValue.apply(Parcel.java:4410)
	at android.os.Parcel$LazyValue.apply(Parcel.java:4369)
	at android.os.BaseBundle.getValueAt(BaseBundle.java:394)
	at android.os.BaseBundle.getValue(BaseBundle.java:374)
	at android.os.BaseBundle.getValue(BaseBundle.java:357)
	at android.os.BaseBundle.getValue(BaseBundle.java:350)
	at android.os.Bundle.getParcelable(Bundle.java:913)
	at androidx.core.os.BundleCompat.getParcelable(BundleCompat.java:73)
	at com.ichi2.anki.services.NotificationService.onReceive(NotificationService.kt:459)
	at android.app.ActivityThread.handleReceiver(ActivityThread.java:4488)
	... 9 more

```

@ericli3690 - could be a weird device; could you take a quick look for anything obvious?

--- a/AnkiDroid/src/main/java/com/ichi2/anki/services/AlarmManagerService.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/services/AlarmManagerService.kt
@@ -271,7 +271,7 @@ class AlarmManagerService : AnkiBroadcastReceiver() {
          * notifications scheduled, this will NOT delete the old scheduled notifications. They must be
          * manually deleted via [unscheduleReviewReminderNotifications].
          */
-        private fun scheduleAllEnabledReviewReminderNotifications(context: Context) {
+        fun scheduleAllEnabledReviewReminderNotifications(context: Context) {
             Timber.d("scheduleAllEnabledReviewReminderNotifications")
             val allReviewRemindersAsMap =
                 ReviewRemindersDatabase.getAllAppWideReminders() + ReviewRemindersDatabase.getAllDeckSpecificReminders()

--- a/AnkiDroid/src/main/java/com/ichi2/anki/services/NotificationService.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/services/NotificationService.kt
@@ -19,6 +19,7 @@ import android.app.PendingIntent
 import android.content.Context
 import android.content.Intent
 import android.graphics.Color
+import android.os.BadParcelableException
 import androidx.annotation.VisibleForTesting
 import androidx.core.app.NotificationCompat
 import androidx.core.app.PendingIntentCompat
@@ -458,11 +459,21 @@ class NotificationService : AnkiBroadcastReceiver() {
             val action = intent.action ?: return
             val extras = intent.extras ?: return
             val reviewReminder =
-                BundleCompat.getParcelable(
-                    extras,
-                    EXTRA_REVIEW_REMINDER,
-                    ReviewReminder::class.java,
-                ) ?: return
+                try {
+                    BundleCompat.getParcelable(
+                        extras,
+                        EXTRA_REVIEW_REMINDER,
+                        ReviewReminder::class.java,
+                    ) ?: return
+                } catch (e: BadParcelableException) {
+                    // #20782: If the extra's review reminder has an invalid schema, attempt a full re-schedule of all
+                    // review reminder notifications, as that will retrieve reminders from Shared Preferences
+                    // and handle any necessary schema updates. It will also trigger immediate notifications
+                    // (and thus this onReceiveBroadcast again for an uncorrupted version of this corrupted review reminder) as necessary.
+                    Timber.w(e, "Failed to get review reminder from intent extras, triggering full re-schedule")
+                    AlarmManagerService.scheduleAllEnabledReviewReminderNotifications(context)
+                    return
+                }
             Timber.d("onReceiveBroadcast: ${reviewReminder.id}")
 
             // We must run some suspending functions. Hence we mark this onReceiveBroadcast function as long-running

